src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosDbCollectionPhysicalPartitionInfo.cs

```diff
@@ -0,0 +1,166 @@
+    /// </summary>
+    internal class CosmosDbCollectionPhysicalPartitionInfo : IRequireInitializationOnFirstRequest, IAsyncDisposable, ICosmosDbCollectionPhysicalPartitionInfo
+    {
+        private readonly CosmosDataStoreConfiguration _dataStoreConfiguration;
+        private readonly CosmosCollectionConfiguration _collectionConfiguration;
+        private readonly IHttpClientFactory _httpClientFactory;
+        private readonly ILogger<CosmosDbCollectionPhysicalPartitionInfo> _logger;
+        private readonly CancellationTokenSource _backgroundLoopCancellationTokenSource = new();
+        private Task _backgroundLoopTask;
+
+        public CosmosDbCollectionPhysicalPartitionInfo(
+            CosmosDataStoreConfiguration dataStoreConfiguration,
+            IOptionsMonitor<CosmosCollectionConfiguration> collectionConfiguration,
+            IHttpClientFactory httpClientFactory,
+            ILogger<CosmosDbCollectionPhysicalPartitionInfo> logger)
+        {
+            EnsureArg.IsNotNull(dataStoreConfiguration, nameof(dataStoreConfiguration));
+            EnsureArg.IsNotNull(collectionConfiguration, nameof(collectionConfiguration));
+            EnsureArg.IsNotNull(httpClientFactory, nameof(httpClientFactory));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _dataStoreConfiguration = dataStoreConfiguration;
+            _collectionConfiguration = collectionConfiguration.Get(Constants.CollectionConfigurationName);
+            _httpClientFactory = httpClientFactory;
+            _logger = logger;
+        }
+
+        public int PhysicalPartitionCount { get; private set; }
+
+        public async Task EnsureInitialized()
+        {
+            PhysicalPartitionCount = await GetPhysicalPartitionCount(_backgroundLoopCancellationTokenSource.Token);
+            _backgroundLoopTask = BackgroundLoop(_backgroundLoopCancellationTokenSource.Token);
+        }
+
+        public async Task BackgroundLoop(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                try
+                {
+                    await Task.Delay(TimeSpan.FromMinutes(10), cancellationToken);
+                    PhysicalPartitionCount = await GetPhysicalPartitionCount(cancellationToken);
+                }
+                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+                {
+                    return;
+                }
+                catch (Exception e)
+                {
+                    _logger.LogError("Unable to get physical partition count.", e);
+                }
+            }
+        }
+
+        private async Task<int> GetPhysicalPartitionCount(CancellationToken cancellationToken)
+        {
+            using HttpClient client = _httpClientFactory.CreateClient();
+
+            string host = _dataStoreConfiguration.Host;
+            string key = _dataStoreConfiguration.Key;
```

Comment:
Does this rotate or change in the managed service?

Reply:
Yes, these change.